### PR TITLE
Support Injector instances in Registry and add inject decorator

### LINKED ISSUE #648: External data property and children injection
**Enhancement**

State injection is currently supported using the `Container`, a HOC pattern to inject a non presentational widget into the tree that works well for simply injecting state/functions etc but does not work nicely when a mixin needs to inject state. Currently `ThemeableMixin` supports "self" injection of state by leveraging the `@beforeRender` and fully replacing the render function with a `w` call for the `Injector` in the registry passing the `render` function as a property.

The main issue with this is that now the result from the widget that used `ThemeableMixin` is actually returned from the `Injector` render function so all the `DNode`s run through the injectors `afterRenders` and not the original widgets. Same for the registries, the registries used will be the registries available to the `Injector` and not the ones provided to the original widget.

This lead to some "hacky" changes within the `Injector` to try and ensure that the correct scope was bound to functions as they were called or for when they would be called in the future.

It would be nice if we could support both styles of external state injection without the need for these "workarounds" and should ensure that the additional workarounds are not required as features are added in the future.

We could potentially could add core support for injection from a context that is stored in the registry (that does not extend `WidgetBase`) within `WidgetBase` using the decorator pattern. `WidgetBase` will run through the property injection and diff'ing internally each render.

```ts
@injector({
    name: 'registry-label',
    getProperties: (context: any, properties: any) => { return properties; }
})
class MyClass extends WidgetBase { }
```

or via the constructor using `addDecorator`.

This should give the building blocks to create the very similar mechanisms available today with `Container`/`Injector` (the container would become a true mixin again) but also support use cases where `mixins` need to inject properties from the registry. With all the required typing goodness too :)

*We would not advocate using the injector standalone as above, instead provide out of the box mixins as well as give all the details needed to create their own `mixin`, that provides all the typings/usability/ergonomics that that users expect*

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Adds `inject` decorator to support side loading properties from a Injector instance registered in the `Registry`.

Includes:
 * Adds additional API for define, get and has that deal with Injector instances (`defineInjector`, `getInjector`, `hasInjector`)
 * Add `inject` decorator that registers a function that will be called with the injector instances context and the widget properties in the `beforeProperties` phase.
 * Updates `Container` & `Themeable` to use mixin pattern for adding `inject` decorator
 * Removes `BaseInjector` that extends `WidgetBase` the `Injector` HOC
 * Renames `WidgetBase.getRegistries()` to `WidgetBase.registries`

Resolves #648

### CODE REVIEW

File: src/Injector.ts [hunk 1]
```diff
@@ -1,154 +1,22 @@
...
-/**
- * Base context class that extends Evented and
- * returns the context using `.get()`.
- */
-export class Context<T = any> extends Evented {
-
-	private _context: T;
-
-	constructor(context: T = <T> {}) {
+	constructor(context: any) {
```
Comment (kitsonk) [inline src/Injector.ts]: What was the reason/motivation for the context being untyped?  Not challenging, just wondering.

File: src/Registry.ts [hunk 2]
```diff
@@ -29,31 +30,55 @@ export interface RegistryEvents extends BaseEventedEvents {
 /**
  * Widget Registry Interface
  */
-export interface Registry {
+export interface RegistryInterface {
```
Comment (kitsonk) [inline src/Registry.ts]: We have traditionally not mentioned `Interface` in our interfaces.  Why the change?

Comment (kitsonk) [reply src/Registry.ts]: Did it have something to do with name merging?  Could this be a use case for an abstract class and having an `AbstractRegistry` as an export?

File: src/decorators/inject.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+export interface GetProperties<T = any> {
+	(context: any, properties: T): T;
+}
+
+export interface InjectConfig {
+	name: RegistryLabel;
+	getProperties: GetProperties;
+}
+
+export function inject({ name, getProperties }: InjectConfig) {
```
Comment (kitsonk) [inline src/decorators/inject.ts]: Some basic JSDoc?

File: src/Registry.ts (hunk 2 above)
Comment (agubler) [reply src/Registry.ts]: The reason was, that without a different name for the interface TS didn't actually error if the class didn't implement it! We've used a suffix of `Interface` in some of the interfaces exported from `interfaces.d.ts` which was the motivation.

I don't think `Abstract` works because the registry isn't actually abstract?

Comment (kitsonk) [reply src/Registry.ts]: Makes sense.  Thanks.

File: src/Injector.ts (hunk 1 above)
Comment (agubler) [reply src/Injector.ts]: There was a challenge with the typing, I'll try it again and let you know (can't remember off the top of my head).

Also wasn't sure what the benefit really was, seen as it's injected at run time there's it didn't really give much to the consumer. The important typing is really on the `getProperties` function that gets passed the context, which the user can type as needed.

File: src/decorators/inject.ts (hunk 3 above)
Comment (agubler) [reply src/decorators/inject.ts]: Ooops yup!

File: src/Injector.ts (hunk 1 above)
Comment (agubler) [reply src/Injector.ts]: I was imaging the challenge it seems, updated to accept a generic type.

Comment (kitsonk) [reply src/Injector.ts]: I do get the point about the _value_ of it...

File: src/Registry.ts [hunk 4]
```diff
@@ -129,18 +171,37 @@ export class Registry extends Evented implements Registry {
...
 			throw error;
 		});
 
 		return null;
 	}
+
+	public getInjector<T extends Injector>(label: RegistryLabel): T | null {
+		if (!this.hasInjector(label)) {
+			if (this.has(label)) {
+				console.warn(`Unable to find injector '${label.toString()}', did you mean to get a widget?`);
```
Comment (matt-gadd) [inline src/Registry.ts]: i'm not convinced we need these warns, because the api is separate and returns discreet types (that extend Injector), the user thus shouldn't be able to get this wrong at design time.

File: src/decorators/inject.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * Decorator retrieves a context from an available registry using the name and
+ * returns calls the `getProperties` function with the existing found context
+ * and passed properties to return the injected properties.
+ *
+ * @param InjectConfig the inject configuration
+ */
+export function inject({ name, getProperties }: InjectConfig) {
+	return handleDecorator((target, propertyKey) => {
+		beforeProperties(function(this: WidgetBase, properties: any) {
+			const context = this.registries.getInjector(name);
```
Comment (matt-gadd) [inline src/decorators/inject.ts]: should this be called `injector`, over `context`?

File: src/decorators/inject.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * Map of instances against registered injectors.
+ */
+const registeredInjectorsMap: WeakMap<WidgetBase, any[]> = new WeakMap();
+
+/**
+ * Defines the contract requires for the get properties function
+ * used to map the injected properties.
+ */
+export interface GetProperties<T = any> {
+	(context: any, properties: T): T;
```
Comment (matt-gadd) [inline src/decorators/inject.ts]: is this `context`?

Comment (agubler) [reply src/decorators/inject.ts]: this is the `context` from the `Injector`

File: src/Container.ts [hunk 7]
```diff
@@ -1,40 +1,26 @@
...
 ): Container<W> {
-	const {
-		getProperties = defaultMappers.getProperties,
-		getChildren = defaultMappers.getChildren
-	} = mappers;
-
-	return class extends WidgetBase<any> {
+	@inject({ name, getProperties })
+	class WidgetContainer extends WidgetBase<Partial<W['properties']>> {
+		public __setProperties__(properties: Partial<W['properties']>): void {
```
Comment (matt-gadd) [inline src/Container.ts]: have we thought how we can better support this in the future? this is the equivalent of `shouldComponentUpdate` set to `true` in react?

File: src/decorators/inject.ts (hunk 5 above)
Comment (agubler) [reply src/decorators/inject.ts]: yup good shout

File: src/decorators/inject.ts (hunk 6 above)
Comment (matt-gadd) [reply src/decorators/inject.ts]: what is a `context`? can we get a better name?

File: src/WidgetBase.ts [hunk 8]
```diff
@@ -706,7 +706,7 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
 			let child: WidgetBaseInterface<WidgetProperties>;
 
 			if (!isWidgetBaseConstructor(widgetConstructor)) {
-				const item = this.getRegistries().get(widgetConstructor);
+				const item = this.registries.get(widgetConstructor);
```
Comment (matt-gadd) [inline src/WidgetBase.ts]: is this the only place the `registries` are accessed? just noticed that we've renamed it back now.

File: src/Container.ts (hunk 7 above)
Comment (agubler) [reply src/Container.ts]: I have thought about it, shall raise an enhancement issue - for now this is the only way I could support it.

File: src/decorators/inject.ts (hunk 6 above)
Comment (agubler) [reply src/decorators/inject.ts]: I am receptive to better names.

File: src/WidgetBase.ts (hunk 8 above)
Comment (agubler) [reply src/WidgetBase.ts]: I think everywhere else actually accesses the private (`_registries`) in `WidgetBase`, I'll update this to do that.

File: src/decorators/inject.ts (hunk 6 above)
Comment (matt-gadd) [reply src/decorators/inject.ts]: I can't participate in suggesting a name without knowing what it is 😄

Comment (agubler) [reply src/decorators/inject.ts]: It's anything that you want to inject, it is what gets passed to the `getProperties` function that returns the values that get injected. It was referred to as a context in the previous implementation also.

```ts
const injector = new Injector(myThingToInject);

const myThingToInject = injector.get();
```

Comment (matt-gadd) [reply src/decorators/inject.ts]: injecting a `payload`?

Comment (agubler) [reply src/decorators/inject.ts]: If you want it to be payload, I don't have any strong attachment to `context` or objections to `payload` - they both work.

Comment (matt-gadd) [reply src/decorators/inject.ts]: the reason i don't like `context` is the baggage that comes from react with it

Comment (agubler) [reply src/decorators/inject.ts]: Yup I understand that. Let me update that for you - I think both are okay and if using `context` could lead to confusing that is it enough reason to change.

Comment (kitsonk) [reply src/decorators/inject.ts]: `value`?  Since you are injecting a _value_ of some sort that can be `any`.  Do we actually expect it to be `any` or is it an `object`?

Comment (agubler) [reply src/decorators/inject.ts]: Aye it can be a `any`, just as the `payload` of an `Injector` can be.
